### PR TITLE
Add return_direct option to tool_calling_agent for direct tool responses

### LINKED ISSUE #758: Allow tool calls to bypass LLM interpretation and return response directly
### Is this a new feature, an improvement, or a change to existing functionality?

New Feature

### How would you describe the priority of this feature request

High

### Please provide a clear description of problem this feature solves

For certain tool calls, I just want to return the result directly without further LLM processing. Currently the Tool Calling Agent always performs LLM processing on the response. 

For example, my tool may output a relatively large dataset that I want to show to the user without any changes or interpretation. Passing the dataset back through the LLM adds significant latency and consumes tokens unnecessarily. 

Even for simple tools like the calculator example, if a user asks "what is 6*7?" I would prefer to immediately return "The product of 6 and 7 is 42" rather than have a slower LLM response simply reword it to say "The answer is 42".

### Describe your ideal solution

A new `return_direct` setting for `tool_calling_agent` that lets you specify which tools (listed under `tool_names`) should bypass LLM processing and return their output directly

In below example, the `calculator_*` tools will return their results directly, while responses from the `fetch_history` and `current_date` tools will be processed by the agent LLM before returning the final answer.
```
workflow:
  _type: tool_calling_agent
  tool_names:
    - calculator_multiply
    - calculator_divide
    - calculator_subtract
    - fetch_history
    - current_date
  llm_name: nim_llm
  verbose: true
  return_direct:
    - calculator_multiply
    - calculator_divide
    - calculator_subtract
```

### Additional context

This direct return behavior conflicts with ReAct's reasoning loop, and is therefore only applicable to `tool_calling_agent`

This behavior is supported in other frameworks like LangChain and LlamaIndex 
https://langchain-ai.github.io/langgraphjs/how-tos/dynamically-returning-directly/
https://docs.llamaindex.ai/en/stable/examples/agent/return_direct_agent/

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct
- [x] I have searched the [open feature requests](https://github.com/NVIDIA/NeMo-Agent-Toolkit/issues?q=is%3Aopen+is%3Aissue+label%3A%22feature+request%22%2Cimprovement%2Cenhancement) and have found no duplicates for this feature request

### PR DESCRIPTION
**Pull request by thepatrickchin**
## Description

- Added optional return_direct: list[str] parameter to constructor, allowing specified tools to return their results directly to the user without additional LLM processing.
- Implemented tool_conditional_edge() method to determine when tools should return directly
- Builds modified state graph following the established pattern from ReWOO agent

Closes #758 

## By Submitting this PR I confirm:
- I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/NeMo-Agent-Toolkit/blob/develop/docs/source/resources/contributing.md).
- We require that all contributors "sign-off" on their commits. This certifies that the contribution is your original work, or you have rights to submit it under the same license, or a compatible license.
  - Any contribution which contains commits that are not Signed-Off will not be accepted.
- When the PR is ready for review, new or existing tests cover these changes.
- When the PR is ready for review, the documentation is up to date with these changes.

## Summary by CodeRabbit

* **New Features**
  * Added an optional return_direct setting to return outputs from specified tools directly to users, enabling early termination of the tool path and faster responses.

* **Documentation**
  * Workflow docs updated to describe the new return_direct option alongside existing configuration settings.

* **Tests**
  * Expanded tests to cover return_direct routing, match/no-match edge cases, and compatibility with existing flows.

### CODE REVIEW

File: src/nat/agent/tool_calling_agent/register.py [hunk 1]
```diff
@@ -47,6 +47,8 @@ class ToolCallAgentWorkflowConfig(FunctionBaseConfig, name="tool_calling_agent")
     system_prompt: str | None = Field(default=None, description="Provides the system prompt to use with the agent.")
     additional_instructions: str | None = Field(default=None,
                                                 description="Additional instructions appended to the system prompt.")
+    return_direct: list[str] | None = Field(
```
Comment (willkill07) [inline src/nat/agent/tool_calling_agent/register.py]: nit: prefer FunctionRef over str for better intent. Or offer a list of Patterns
```suggestion
    return_direct: list[FunctionRef] | None = Field(
```

File: src/nat/agent/tool_calling_agent/agent.py [hunk 2]
```diff
@@ -146,13 +151,60 @@ async def tool_node(self, state: ToolCallAgentGraphState):
             logger.error("%s Failed to call tool_node: %s", AGENT_LOG_PREFIX, ex)
             raise
 
-    async def build_graph(self):
+    async def tool_conditional_edge(self, state: ToolCallAgentGraphState) -> AgentDecision:
```
Comment (willkill07) [inline src/nat/agent/tool_calling_agent/agent.py:154]: Add docstring for function

File: src/nat/agent/tool_calling_agent/agent.py [hunk 3]
```diff
@@ -57,12 +60,14 @@ def __init__(
         detailed_logs: bool = False,
         log_response_max_chars: int = 1000,
         handle_tool_errors: bool = True,
+        return_direct: list[str] | None = None,
```
Comment (willkill07) [inline src/nat/agent/tool_calling_agent/agent.py]: nit: be type safe or -- alternatively, permit a list of Patterns for matching.
```suggestion
        return_direct: list[BaseTool] | None = None,
```

File: src/nat/agent/tool_calling_agent/agent.py [hunk 4]
```diff
@@ -85,8 +90,8 @@ def __init__(
             )
 
         self.agent = prompt_runnable | self.bound_llm
-
         self.tool_caller = ToolNode(tools, handle_tool_errors=handle_tool_errors)
+        self.return_direct = return_direct
```
Comment (willkill07) [inline src/nat/agent/tool_calling_agent/agent.py]: Then here you can map the BaseTool's name to the list for reuse.
e.g. `[tool.name for tool in return_direct]`

File: src/nat/agent/tool_calling_agent/agent.py [hunk 5]
```diff
@@ -146,13 +151,60 @@ async def tool_node(self, state: ToolCallAgentGraphState):
...
-            )
+            logger.debug("%s Starting the Tool Conditional Edge", AGENT_LOG_PREFIX)
+            if not state.messages:
+                logger.debug("%s No messages in state; routing to agent", AGENT_LOG_PREFIX)
+                return AgentDecision.TOOL
+
+            last_message = state.messages[-1]
+            # Return directly if this tool is in the return_direct set
+            if (self.return_direct and isinstance(last_message, ToolMessage) and last_message.name
+                    and last_message.name in self.return_direct):
```
Comment (willkill07) [inline src/nat/agent/tool_calling_agent/agent.py:174]: If you do switch to a list of patterns, update logic here

**Comment by willkill07** [comment]
/ok to test [129fc03](https://github.com/NVIDIA/NeMo-Agent-Toolkit/pull/775/commits/129fc03902eaaa6335d4f5f80e00c5e7e2f56a10)

File: src/nat/agent/tool_calling_agent/agent.py [hunk 6]
```diff
@@ -19,10 +19,13 @@
 from langchain_core.callbacks.base import AsyncCallbackHandler
 from langchain_core.language_models import BaseChatModel
 from langchain_core.messages import SystemMessage
+from langchain_core.messages import ToolMessage
 from langchain_core.messages.base import BaseMessage
 from langchain_core.runnables import RunnableLambda
 from langchain_core.runnables.config import RunnableConfig
 from langchain_core.tools import BaseTool
+from langgraph.graph import StateGraph
+from langgraph.graph.graph import CompiledGraph
```
Comment (willkill07) [inline src/nat/agent/tool_calling_agent/agent.py]: I just merged a PR that bumps the version of LangGraph.

This should be
```suggestion
from langgraph.graph.state import CompiledStateGraph
```

File: src/nat/agent/tool_calling_agent/agent.py [hunk 7]
```diff
@@ -146,13 +151,60 @@ async def tool_node(self, state: ToolCallAgentGraphState):
...
+            else:
+                # Continue to agent for processing
+                logger.debug("%s Tool response will be processed by agent", AGENT_LOG_PREFIX)
+                return AgentDecision.TOOL
+        except Exception as ex:
+            logger.exception("%s Failed to determine tool conditional edge: %s", AGENT_LOG_PREFIX, ex)
+            logger.warning("%s Continuing to agent for processing", AGENT_LOG_PREFIX)
+            return AgentDecision.TOOL
+
+    async def _build_graph(self, state_schema) -> CompiledGraph:
```
Comment (willkill07) [inline src/nat/agent/tool_calling_agent/agent.py]:
```suggestion
    async def _build_graph(self, state_schema) -> CompiledStateGraph:
```

File: src/nat/agent/tool_calling_agent/agent.py [hunk 8]
```diff
@@ -146,13 +151,60 @@ async def tool_node(self, state: ToolCallAgentGraphState):
...
+        try:
+            logger.debug("%s Building and compiling the Tool Calling Agent Graph", AGENT_LOG_PREFIX)
+
+            graph = StateGraph(state_schema)
```
Comment (willkill07) [inline src/nat/agent/tool_calling_agent/agent.py:181]:
```suggestion
            graph = CompiledStateGraph(state_schema)
```

File: src/nat/agent/tool_calling_agent/agent.py [hunk 9]
```diff
@@ -146,13 +151,60 @@ async def tool_node(self, state: ToolCallAgentGraphState):
...
+
+            graph.set_entry_point("agent")
+            self.graph = graph.compile()
+
+            return self.graph
+        except Exception as ex:
+            logger.error("%s Failed to build Tool Calling Agent Graph: %s", AGENT_LOG_PREFIX, ex)
+            raise
+
+    async def build_graph(self) -> CompiledGraph:
```
Comment (willkill07) [inline src/nat/agent/tool_calling_agent/agent.py]:
```suggestion
    async def build_graph(self) -> CompiledStateGraph:
```

File: src/nat/agent/tool_calling_agent/agent.py (hunk 6 above)
Comment (thepatrickchin) [reply src/nat/agent/tool_calling_agent/agent.py]: I've rebased my branch and updated these imports

File: src/nat/agent/tool_calling_agent/agent.py (hunk 3 above)
Comment (thepatrickchin) [reply src/nat/agent/tool_calling_agent/agent.py]: @willkill07 I've updated the type to `BaseTool`. Regarding your pattern matching suggestion, do you mean supporting something like this so all calculator tools will return directly?
```
workflow:
  ...
  return_direct:
    - calculator_*
```

Comment (willkill07) [reply src/nat/agent/tool_calling_agent/agent.py]: It could be defined as a glob (shown) or a regex (`calculator.*`), but essentially, yes.

File: src/nat/agent/tool_calling_agent/agent.py (hunk 4 above)
Comment (willkill07) [inline src/nat/agent/tool_calling_agent/agent.py]: If we don't internally map to tool names, then the comparison/check will always fail
```suggestion
        self.return_direct = [tool.name for tool in return_direct]
```

**Comment by willkill07** [comment]
/ok to test [36a4859](https://github.com/NVIDIA/NeMo-Agent-Toolkit/pull/775/commits/36a4859a78178ab9fcc49f50ff8f3c8c72b466c3)

**Review by willkill07 — APPROVED** [review]
